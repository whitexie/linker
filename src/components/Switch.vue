<script setup lang="ts">
interface Props {
  modelValue: boolean | string | number
  size?: number
  activeColor?: string
  inactiveColor?: string
  activeValue?: boolean | string | number
  inactiveValue?: boolean | string | number
}

const props = withDefaults(defineProps<Props>(), {
  size: 16,
  activeColor: '#30a72d',
  inactiveColor: '#676b5f',
  activeValue: true,
  inactiveValue: false,
})
const emits = defineEmits<{
  (e: 'update:modelValue', value: boolean | string | number): void
}>()

const isOpen = computed(() => {
  return props.modelValue === props.activeValue
})

const switchStyle = computed(() => {
  return {
    'width': `${props.size * 2 + 2}px`,
    'height': `${props.size + 4}px`,
    'background-color': isOpen.value ? props.activeColor : props.inactiveColor,
    'border-radius': `${props.size + 4}px`,
    'overflow': 'hidden',
  }
})
const dotStyle = computed(() => {
  return {
    width: `${props.size}px`,
    height: `${props.size}px`,
    transform: isOpen.value ? 'translateX(100%)' : 'translateX(0)',
  }
})
function handleClick() {
  let value = props.activeValue
  if (props.modelValue === props.activeValue)
    value = props.inactiveValue

  emits('update:modelValue', value)
}
</script>

<template>
  <div flex items-center cursor-pointer :style="switchStyle" @click="handleClick">
    <div transition-transform-200 bg-white class="rounded-50%" :style="dotStyle" />
  </div>
</template>

<style lang="scss" scoped>
</style>
