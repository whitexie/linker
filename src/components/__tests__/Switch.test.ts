import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import Switch from '../Switch.vue'

describe('Counter.vue', () => {
  it('should render', () => {
    const wrapper = mount(Switch, { props: { modelValue: false } })
    expect(wrapper.find('.dot').exists()).toBe(true)
    expect(wrapper.find('.dot').attributes('class')).toContain('rounded-50%')
    // expect(wrapper.html()).toMatchSnapshot()
  })

  it('should be interactive', async () => {
    const wrapper = mount(Switch, { props: { modelValue: false } })

    await wrapper.find('.switch').trigger('click')
    // console.log('wrapper.emitted() => ', wrapper.emitted())
    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([true])
  })
})
