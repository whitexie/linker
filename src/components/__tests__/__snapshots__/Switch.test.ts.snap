// Vitest Snapshot v1

exports[`Counter.vue > should render 1`] = `
"<div flex=\\"\\" items-center=\\"\\" cursor-pointer=\\"\\" style=\\"width: 34px; height: 20px; background-color: rgb(103, 107, 95); border-radius: 20px; overflow: hidden;\\">
  <div transition-transform-200=\\"\\" bg-white=\\"\\" class=\\"dot rounded-50%\\" style=\\"width: 16px; height: 16px; transform: translateX(0);\\"></div>
</div>"
`;
